README.md
macroutils
==========

Page content: 
[In short](#inshort) &middot; 
[Installation](#installation) &middot; 
[Official CKB release](#official) &middot; 
[Development version](#development) &middot; 
[Text based graphical user interface](#text_gui) &middot; 
[Documentation](#doc) &middot; 
[Issues and bug reports](#issues) 



_R utility functions for the [MACRO][] and [SOILNDB][] models._ 
_Read and write binary files, create plots, and more._

[MACRO][] and [SOILNDB][] use [binary files][binary_files] for 
both input (climate variables) and output (simulation results). 
The format of these files is not standard and `macroutils` 
provides functions to handle them. 



In short    <a id="inshort"></a>
========

*   Author: **[Julien MOEYS][julienmoeys]** ([SLU][]/[CKB][])
*   [Tutorial][macroutils_tuto] and [On-line documentation][macroutils_help] 
*   Package description: See [DESCRIPTION](https://github.com/julienmoeys/macroutils/blob/master/pkg/macroutils/DESCRIPTION).
*   View and report issues: https://github.com/julienmoeys/macroutils/issues 
    Note: _not_ for reporting issues related to MACRO (core 
    model)
*   Homepage of the MACRO model: [here][MACRO]



Installation    <a id="installation"></a>
============

For all versions you need to install beforehand the sofware [R][] 
("a free software environment for statistical computing and graphics").

From R homepage, go to the [CRAN homepage][CRAN] or (better) choose 
the nearest [CRAN mirror][CRAN_mirrors].

From there, choose 'Download R for Windows' (or Mac or Linux), and 
then choose 'base' or 'install R for the First time'.

*   If you want to install the official CKB release (see below), you 
    must download the R version that corresponds to the package 
    requirement as indicated on `macroutils` / MACRO 5.2 homepage 
    on [CKB][] website (see below).

*   If you want to install the development version of macroutils (see 
    below), you can download latest R version, or (safer) the R 
    version that is indicated in the 
    [DESCRIPTION](/pkg/macroutils/DESCRIPTION) file.

Download the R installer, and install [R][] (follow the 
instructions).



Official [CKB][] release    <a id="official"></a>
------------------------

Note: _Windows only_ (the source package is not provided on this 
page).

On [MACRO 5.2 official homepage][MACRO], you can download the 
latest version of the package

The package binary Files are usually named `macroutils_x.y.z.zip`, 
where `x`, `y` and `z` form the version number (major milestone, 
minor milestone, minor revision).

Notice that you don't need to uncompress this `.zip` File! (it is 
an installer). Open R Graphical User Interface, and in the packages
menu (top menu bar), choose 'Install package(s) from local zip 
File(s)', and select the package binary File (`macroutils_x.y.z.zip`).



Development version    <a id="development"></a>
-------------------

Note: _May work on other platforms than Windows, but this has not_ 
_been tested_ ([MACRO][] only runs on Windows).

Note: _If you are not familiar with R (and R developments), you 
should probably avoid that method, as some complications may arise_.

The development version of `macroutils` is publicly available 
on [GitHub][] ([the page you are actually reading][macroutils_gh]). 
This page allows you to explore the package source code, to install 
the package development version (see below) and to report some 
[issues][macroutils_issues] (bugs)

*   On Windows computers, install [Rtools][] Install the 
    version that corresponds to your the version of R that is 
    installed on your computer. On other machines, read the 
    [instructions here][devtools_readme] (Windows users may also 
    have a look at this page).
    
*   Install the package [devtools][]. `devtools` is a package 
    developed by [Hadley Wickham][HadleyWickham] to install 
    development version of R packages that are hosted on [GitHub][] 
    (among other useful things).
    
*   Open a new R cession and type `install.packages("devtools")`.
    Follow the additional [instructions here][devtools_readme].
    
*   Install the development version of `macroutils`. Open a new R 
    cession and type:


    devtools::install_github("julienmoeys/macroutils/pkg/macroutils")



Text based graphical user interface    <a id="text_gui"></a>
===================================

See also: The [package tutorial][macroutils_tuto]. 

`macroutils` provides simple but effective [text based graphical 
user interfaces][text_gui] for the most useful functions 
provided by the package. Once the package has been 
[installed](#installation), you just need to type one 
simple command (per function) and then follow the instructions 
given to you.



Plotting the content of one (or several) bin file(s)
----------------------------------------------------

Launch R and type:

    macroutils::macroPlot()

and follow the instructions. This is equivalent to:

    library( "macroutils" )
    macroPlot()



Converting one (or several) bin file(s)
----------------------------------------------------

Launch R and type:

    macroutils::macroConvertBin()

and follow the instructions. This is equivalent to:

    library( "macroutils" )
    macroConvertBin()



View one bin file
-----------------

Launch R and type:

    macroutils::macroViewBin()

and follow the instructions. This is equivalent to:

    library( "macroutils" )
    macroViewBin()




Documentation    <a id="doc"></a>
=============

See also: The [package tutorial][macroutils_tuto]. 



On-line documentation
---------------------

The package help pages can be browsed on-line 
[from this page][macroutils_help]. Notice that this may not 
always be the very latest version of the package.



Loading the package
-------------------

Open R graphical user interface and type:

    library( "macroutils" )



Package documentation (help pages)
----------------------------------

To go further, you can access the help page by typing:

    help( package = "macroutils" )

There, see in particular the vignette (tutorial) in "User guides, 
package vignettes and other documentation" (to get started). You 
can also access that tutorial with the command

    vignette( "macroutils_vignette" )

Notice that if you have installed the development version from 
[GitHub][] the vignette may not have been generated with the 
package. But the vignette can be extracted from the package 
Windows binaries that can be downloaded on [MACRO][] homepage 
(`macroutils_x.y.z.zip` > `macroutils` > `doc` > 
`macroutils_vignette.pdf`).



Issues and bug reports    <a id="issues"></a>
======================

Issues and bugs can be reported [here][macroutils_issues]. Do not 
report issues related to the model [MACRO][] (i.e. not related to 
`macroutils`) on that page. Do not either report more general 
questions that are related to [R][] and R usage. Provide a 
reproducible (_self standing_) example of your problem, as it 
generally greatly helps to narrow down the problem.



<!--- Links         -->
[SLU]:              http://www.slu.se/en/ "Swedish University of Agricultural Sciences"
[CKB]:              http://www.slu.se/en/collaborative-centres-and-projects/centre-for-chemical-pesticides-ckb1/ "The Centre for Chemical Pesticides (CKB)"
[MACRO]:            http://www.slu.se/en/collaborative-centres-and-projects/centre-for-chemical-pesticides-ckb1/areas-of-operation-within-ckb/models/macro-52/ "MACRO - pesticide fate in soils (SLU/CKB)"
[GitHub]:           https://github.com/ "GitHub"
[macroutils_gh]:    https://github.com/julienmoeys/macroutils "R package macroutils (on GitHub)"
[macroutils_issues]:https://github.com/julienmoeys/macroutils/issues "Issues on the package macroutils"
[macroutils_help]:  http://julienmoeys.github.io/macroutils/ "Documentation for package 'macroutils'"
[R]:                http://www.r-project.org/ "R is a free software environment for statistical computing and graphics"
[CRAN]:             http://cran.r-project.org/ "The Comprehensive R Archive Network"
[CRAN_mirrors]:     http://cran.r-project.org/mirrors.html "CRAN Mirrors"
[Rtools]:           http://cran.r-project.org/bin/windows/Rtools/ "Rtools: Building R for Windows"
[devtools_readme]:  http://cran.r-project.org/web/packages/devtools/README.html "README page of devtools"
[devtools]:         http://cran.r-project.org/web/packages/devtools "R package devtools"
[HadleyWickham]:    http://had.co.nz/ "Hadley Wickham homepage"
[text_gui]:         http://en.wikipedia.org/wiki/Text-based_user_interface "Text-based user interface (Wikipedia)"
[julienmoeys]:      http://www.julienmoeys.info "Julien Moeys homepage"
[binary_files]:     https://en.wikipedia.org/wiki/Binary_file "Binary files (Wikipedia)"
[SOILNDB]:          http://www.slu.se/en/collaborative-centres-and-projects/slu-water-hub/models/soilndb/ "SoilN DB homepage" 
[macroutils_tuto]:  https://github.com/julienmoeys/macroutils/blob/master/pkg/macroutils/inst/doc/macroutils_vignette.pdf "macroutils tutorial" 



NEWS

    R package: macroutils

CHANGES IN VERSION 1.12.0

    2015/12/08  macroBugFixCleanDb() now issues a warning 
        message if not run in R 32 bit (i386), as RODBC 
        for MS Access works (only?) with R 32 bit.
        
        macroBugFixCleanDb() now also attempts to fix a 
        bug that appear when increasing (changing) the 
        number of numerical layers. It seems some values 
        in the parameter table `OutputLayers` (new numerical 
        layers) have their column `Selected` empty, instead 
        of 0 or 1. The bug fix will also attempt to fix 
        value in `OutputLayers` that are present despite 
        the parameter in `Output()` having selected set 
        to 0.
    
CHANGES IN VERSION 1.11.0
    
    2015/11/02  macroInFocusGWConc() now calculates the 
        mass fraction of the pesticide flow that is 
        due to micropore or/versus macropore transport.
        This experimental feature aims at estimating the 
        importance of macropore flow in the calculated 
        concentration.
    
CHANGES IN VERSION 1.10.1
    
    2015/10/09  macroInFocusGWConc() now checks for negative 
        concentrations in the 80th percentile (or can set 
        negative concentrations to 0)
    
CHANGES IN VERSION 1.10.0
    
    2015/10/09  Implemented more strict tests / benchmarks 
        between macroReadBin() and bin-file conversion 
        performed by MACRO 5.2. Should be able to detect 
        unwanted accidental changes in macroReadBin() 
        (at least in terms of numerical accuracy)
        
        Also included a simple benchmark in macroInFocusGWConc()
        on concentration and percolation.
        
CHANGES IN VERSION 1.9.0
    
    2015/10/08  Added the function macroInFocusGWConc() 
        to calculate pesticide concentrations from MACROInFOCUS 
        output bin file(s). Internal and experimental.
    
CHANGE SIN VERSION 1.8.4
    
    2015/09/23  Added an 'overwrite' agument to 
        macroConvertBin(), and fixed a bug that caused an 
        interactive question to be asked to the user even 
        when 'gui' was FALSE and the output files already 
        existed. Now if 'overwrite' is FALSE and 'gui' is 
        TRUE a question is asked to the user, and otherwise 
        if 'gui' is FALSE and 'overwrite' is FALSE, an 
        error is issued when files already exists.
    
        A sub-section was added in the vignette that describes 
        how to perform a batch conversion of files with 
        macroConvertBin(), including listing all the bin 
        files in a folder.
    
CHANGES IN VERSION 1.8.3

    2015/07/23  Added the function isValidTimeSeries(), used
        internally to check that time series read from or 
        written to BIN-files are consistent (increasing, 
        homogeneous increment and no duplicated values).
    
CHANGES IN VERSION 1.8.2
    
    2015/04/17  Compiled for R 3.2.0 RC (2015-04-15 r68178).
    
        Cleaned-up internal call to library() or require() 
        and use :: instead (+ Depends or Suggests or Import 
        fields in package DESCRIPTION)
    
    2015/05/04  Compiled for R 3.2.0 Official release.
    
CHANGES IN VERSION 1.8.1
    
    2015/03/17  Moved from a local SVN to a public git (GitHub) 
        repository. Last SVN version was 18:52M (macroutils 1.8.0)
    
CHANGES IN VERSION 1.8.0
    
    2015/03/17  From now on macroutils will contain only "official" 
        MACRO routines and utilities, and the routines related 
        to crop growth (under development) have been moved to 
        another R package, macrocrop (not publicly released).
        
        macroutils was split for SVN version "18:50M" (macroutils 
        version 1.7.4).
        
CHANGES IN VERSION 1.7.4
    
    2014/12/08  Bug fix in macroLAI()
    
CHANGES IN VERSION 1.7.3

    2014/09/28  Compiled for R 3.1.1
    
CHANGES IN VERSION 1.7.2

    2014/07/03  Added the function macroutilsInfo() that fetches 
        and return a rather comprehensive report on system info 
        and package(s) version(s) and MD5, either for macroutils 
        or for any other package
        
        SVN revision is now displayed when the package is attached
        
CHANGES IN VERSION 1.7.1
    
    2014/05/28 to 2014/06/03    Attempts to improve the LAI-calculations 
        for multiple crops. Did not work in some cases.
    
CHANGES IN VERSION 1.7.0
    
    2014/05/26  Now macroLAI() can handle a data.frame of 
        crop parameters (multiple crops).
        
        Parameter "Date" in macroLAI() renamed 'x'
        
        macroIntercept() parameters now simplified
    
CHANGES IN VERSION 1.6.2
    
    2014/05/22  fixed a bug in macroLAI() and macroRootDepth() 
        and macroRootDensity(): parameters iharv and laiharv renamed 
        iharv and laihar (now correct)
        
        bug fix: macroPlot() was not always identifying correctly 
            when given a list of data.frame to plot
        
CHANGES IN VERSION 1.6.1
    
    2014/05/22  macroPlot() now includes a parameter 'z' for 
        pre-selecting the variables to include in the graph
    
CHANGES IN VERSION 1.6.0

    2014/05/15  Included macroLAI() and macroIntercept() to calculate 
        total and green LAI a-la-MACRO and the fraction of irrigation 
        water intercepted by the crop vegetation.
    
CHANGES IN VERSION 1.5.0

    2014/05/12  Included macroRootDepth() and macroRootDensity(), 
        functions to estimate root depth and density as in 
        MACRO

CHANGES IN VERSION 1.4.0
    
    2014/05/12  Included a bug-fix for MACRO GUI parameter database, 
        macroBugFixCleanDb(), that remove orphan and duplicated 
        values in the table `Output()`. Might be deleted when 
        the bug has been fixed in MACRO GUI.
    
CHANGES IN VERSION 1.3.0
    
    2014/03/19
        
        macroReadBin() now output macroTimeSeries or 
        macroTimeSeriesList objects (class), instead of 

pkg/macroutils/R/bugFixes.R


# .chooseAccessFiles ============================================

#'@importFrom tcltk tk_choose.files

## # Pop-up a menu to choose MS Access file from the file system.
## # 
## # Pop-up a menu to choose MS Access file from the file system.
## #
## #
## #@param caption
## #   See \code{\link[utils]{choose.files}} or 
## #   \code{\link[tcltk]{tk_choose.files}}.
## # 
## #@param multi
## #   See \code{\link[utils]{choose.files}} or 
## #   \code{\link[tcltk]{tk_choose.files}}.
## # 
## # 
.chooseAccessFiles <- function(
    caption = "Select one or several MACRO parameter database(s) (MS Access)", 
    multi   = TRUE
){  
    if( !interactive() ){ 
        stop( "'.chooseAccessFiles' can only be used in interactive mode" )
    }   
    
    
    ## Set the folder working directory
    lastBinWd <- getMuPar( "lastBinWd" ) 
    
    if( length(lastBinWd) == 0 ){ 
        lastBinWd <- getwd() 
    }else{ 
        if( lastBinWd == "" ){ 
            lastBinWd <- getwd() 
        }else{ 
            lastBinWd <- file.path( lastBinWd, "*.*" )
        }   
    }   
    
    
    ## Create a template of file extension to be read:
    filterz <- matrix( 
        data  = c( 
            "Access files (*.mdb)", "*.mdb", 
            "All",                  "*" ), 
        nrow  = 2, 
        ncol  = 2, 
        byrow = TRUE  
    )   
    rownames( filterz ) <- c( "mdb", "all" ) 
    
    ## Pop-up a menu to choose the bin file to be 
    ## imported
    if( exists(x = "choose.files", where = "package:utils" ) ){ 
        # fun <- get( "choose.files" ) 
        
        file <- utils::choose.files(
            default = lastBinWd, # , "*.bin"
            caption = caption, 
            multi   = multi, 
            filters = filterz 
        )   
        
    }else{ 
        # library( "tcltk" ) 
        
        # fun <- get( "tk_choose.files" ) 
        
        file <- tcltk::tk_choose.files(
            default = lastBinWd, # , "*.bin"
            caption = caption, 
            multi   = multi, 
            filters = filterz 
        )   
    }   
    
    
    # browser()
    
    
    ## Set the last folder where binary files were found:
    lastBinWd <- .pathNoLastItem( p = file[1] ) 
    
    muPar( "lastBinWd" = lastBinWd ) 
    
    return( file ) 
}   



# macroBugFixCleanDb ============================================

#' Clean-up MACRO 5.2 parameter databases. Fixes 4 known bugs (orphan, incomplete or unnecessary values)
#'
#' Clean-up MACRO 5.2 parameter databases. Fixes 4 known bugs 
#'  (orphan, incomplete or unnecessary values). It is very 
#'  highly recommended to make a backup-copy of MACRO 5.2 
#'  parameter databases before you try this utility. The 
#'  R \bold{\code{\link[RODBC]{RODBC-package}}} is required to run 
#'  this function, and you also need to run a \bold{32 bit 
#'  (i386)} version of R (maybe located in 
#'  \code{\{R installation directory\}bin/i386/Rgui.exe}, 
#'  if it has been installed).
#'
#'
#'@param file 
#'  Vector of character strings or a single \code{\link{connection}}
#'  to a MACRO GUI MS Access parameter database. If a vector of 
#'  character strings, it should be the name(s) of
#'  the Access database(s) containing MACRO parameters. The path 
#'  of the file(s) may be provided as well, if file(s) 
#'  is (are) not in the working directory.
#'
#'@param paranoia 
#'  Single logical value. If \code{TRUE}, the user is asked 
#'  if he made a backup copy of the parameter database.
#'
#'@param \dots Additional options passed to specific 
#'  methods.
#'
#'
#'@return 
#'  Do not return anything. Used for it side effect on a MACRO 
#'  parameter database.
#'
#'
#'@rdname macroBugFixCleanDb
#'
#'@export
#'
#'
macroBugFixCleanDb <- function(
 file, 
 paranoia = TRUE, 
 ...
){  
    # paranoia <- TRUE; file <- "D:/Users/julienm/Documents/_WORKS/_PROJECTS/r_packages/perform/pkg/rmacro/inst/PROJECT_MACRO_5.rmacro.mdb"
    
    testRODBC <- "RODBC" %in% rownames( installed.packages() )
    
    if( !testRODBC ){ 
        stop( "'RODBC' package not available. Please install RODBC first: install.package('RODBC')" )
    }else{ 
        arch <- utils::sessionInfo()[[ "R.version" ]][[ "arch" ]]
        
        if( arch != "i386" ){
            warning( sprintf( 
                "'RODBC' MS Access interface requires a 32 bit version of R (i386) (now: %s). Consider running R i386 instead ({R install dir}/i386/Rgui.exe)", 
                arch
            ) ) 
        }   
    }   
    
    
    ## If no file name is provided
    if( missing( file ) ){ 
        if( interactive() ){ 
            ## Pop-up a menu to choose the bin file to be 
            ## imported
            file <- .chooseAccessFiles(
                caption = "Select one or several MACRO parameter database(s) (MS Access)", 
                multi   = TRUE  
            )   
            
            if( length(file) == 0 ){ 
                stop( "You haven't choosen any binary file to read :o(" )
            }   
            
            file <- sort( file ) 
        }else{ 
            stop( "'file' can not be missing when R is not being used interactively" ) 
        }   
    }   
    
    
    if( interactive() & paranoia ){ 
        cp <- select.list( 
            title       = "Did you made a backup-copy of your parameter database?",
            choices     = c( "Yes", "No" ), 
            preselect   = NULL, 
            multiple    = FALSE, 
            graphics    = FALSE 
        )   
        
        if( cp == "No" ){ 
            stop( "Then make a backup-copy of your parameter database" )
        }   
    }   
    
    
    silent <- lapply( 
        X   = file, 
        FUN = function(f){ 
            # f <- file[1]
            
            message( sprintf( "Starts processing database: '%s'.", f ) )
            
            channel <- RODBC::odbcConnectAccess( access.file = f ) 
            
            on.exit( try( RODBC::odbcClose( channel = channel ) ) )
            
            tablesList <- RODBC::sqlTables( channel = channel )
            
            .tables <- c( "Output()", "Run_ID", "OutputLayers" ) 
            testTables <- .tables %in% 
                tablesList[, "TABLE_NAME" ]
            
            if( !all(testTables) ){ 
                stop( sprintf( 
                    "The table(s) %s cannot be found in the database (%s)", 
                    paste( .tables[ !testTables ], collapse = "; " ), 
                    f 
                ) ) 
            };  rm( .tables, testTables ) 
            
            output <- RODBC::sqlFetch( channel = channel, sqtable = "Output()" )
            
            runIdTbl <- RODBC::sqlFetch( channel = channel, sqtable = "Run_ID" )
            
            outputLayers <- RODBC::sqlFetch( channel = channel, sqtable = "OutputLayers" )
            
            # runIds <- runIdTbl[, "R_ID" ]
            
            
            
            # 1 - ORPHAN `R_ID` IN `Output()` (NOT ANY MORE 
            #     IN `Run_ID`)
            # ----------------------------------------------
            
            #   ID in "Output()" but not in "Run_ID"
            missId <- unique( missId0 <- output[ 
                !(output[, "R_ID" ] %in% runIdTbl[, "R_ID" ]), 
                "R_ID" ] ) 
            
            #   Delete IDs in Output() that are 'orphan'
            if( length( missId ) > 0 ){ 
                message( sprintf( 
                    "Found %s orphan values in `Output()` for RUNID(s) %s", 
                    length( missId0 ), 
                    paste( missId, collapse = "; " )
                ) ) 
                
                rm( missId0 )
                
                for( id in missId ){ 
                    RODBC::sqlQuery( 
                        channel = channel, 
                        query   = sprintf( "DELETE * FROM `Output()` WHERE `R_ID` = %s", id ), 
                    )   
                }   
                
                message( "Orphan values deleted in `Output()`" )
            }else{
                message( "Found no orphan values in `Output()` (fine!)" )
            }   
            
            rm( missId )
            
            #   Re-fetch Output()
            output <- RODBC::sqlFetch( channel = channel, sqtable = "Output()" ) 
            
            
            
            # 2 - DUPLICATED `R_ID`-`Var` IN `Output()`
            # ----------------------------------------------
            
            #   Find RUNID with duplicated export parameters
            uOutput   <- output[, c( "R_ID", "Var" ) ] 
            selDuplic <- duplicated( uOutput ) 
            
            duplicId  <- unique( uOutput[ selDuplic, "R_ID" ] ) 
            rm( uOutput, selDuplic )
            
            if( length( duplicId ) > 0 ){ 
                message( sprintf( 
                    "Found %s duplicated values in `Output()` for RUNID(s) %s", 
                    length( selDuplic ), 
                    paste( duplicId, collapse = "; " ) 
                ) ) 
                
                for( id in duplicId ){ 
                    # id <- duplicId[ 1 ]
                    
                    sOutput <- subset( 
                        x      = output, 
                        subset = eval( quote( R_ID == id ) ) )
                    
                    #   Order the table
                    sOutput <- sOutput[ 
                        order( sOutput[, "Var" ], sOutput[, "Output()ID" ] ), ]
                    
                    #   Unique list of variables
                    uVar <- unique( sOutput[, "Var" ] ) 
                    
                    nrow( sOutput ) # 98
                    
                    for( v in uVar ){ 
                        # v <- uVar[ 1 ]
                        
                        outputId <- sOutput[ 
                            sOutput[,"Var"] == v, 
                            "Output()ID" ] 
                        
                        if( length( outputId ) > 1 ){ 
                            RODBC::sqlQuery( 
                                channel = channel, 
                                query   = sprintf( "DELETE * FROM `Output()` WHERE `Output()ID` = %s", min( outputId ) ), 
                            )   
                        }   
                        
                        rm( outputId )
                        
                    }   
                    
                    # nrow( sOutput ) # 98
                    
                    rm( sOutput, uVar, v )
                }   
                
                message( "Duplicated values deleted in `Output()`" )
            }else{
                message( "Found no duplicated values in `Output()` (fine!)" )
            }   
            
            rm( duplicId )
            
            
            
            # 3 - EXPORT PARAMS IN `Output()` NOT SELECTED 
            #     BUT STILL PRESENT IN `OutputLayers`
            # ----------------------------------------------
            
            #   Find outputs that are not selected in `Output()`
            #   but nonetheless present in `OutputLayers`
            uOutput2 <- unique( output[, c( "R_ID", "Var", "Output()ID", "selected" ) ] )
            
            #   Only keep those that are not selected 
            #   as layered output
            uOutput2 <- subset( x = uOutput2, subset = eval( quote( selected != 1 ) ) )
            
            #   Find the one that should not be there
            testOutLay <- outputLayers[, "Output()ID" ] %in% 
                uOutput2[, "Output()ID" ]
            
            if( any( testOutLay ) ){
                #   Reverse selection: entries in `Output()`
                #   that have unnecessary layers parameters in 
                #   `OutputLayers`
                testOut <- uOutput2[, "Output()ID" ] %in% 
                    outputLayers[ testOutLay, "Output()ID" ] 
                
                message( sprintf( 
                    "Found %s unnecessary entries in `OutputLayers` for RUNID(s) %s", 
                    length( testOutLay ),
                    paste( unique( uOutput2[ testOut, "R_ID" ] ), collapse = "; " ) 
                ) ) 
                
                rm( testOut )
                
                #   Find the OutputLayerID to be removed
                idOut <- outputLayers[ testOutLay, "OutputLayerID" ]
                
                RODBC::sqlQuery( 
                    channel = channel, 
                    query   = sprintf( 
                        "DELETE * FROM `OutputLayers` WHERE `OutputLayerID` IN (%s)", 
                        paste( as.character( idOut ), collapse = ", " ) 
                    ),  
                )   
                
                message( sprintf( 
                    "Deleted %s unnecessary entries in `OutputLayers`", 
                    length( idOut ) 
                ) ) 
                
                rm( idOut )
            }else{
                message( "Found no unnecessary entries in `OutputLayers` (fine!)" )
            }   
            
            rm( uOutput2, testOutLay )
            
            
            
            # 4 - EXPORT PARAMS IN `OutputLayers` WHERE THE 
            #     COLUMN `Selected` IS NOT SET (neither 0 nor 
            #     1), presumably after more layers were 
            #     added
            # ----------------------------------------------
            
            uOutput2 <- unique( output[, c( "R_ID", "Var", "Output()ID", "selected" ) ] )
            
            #   Find the one that should not be there
            selFixSelCol <- is.na( outputLayers[, "Selected" ] )
            
            if( any( selFixSelCol ) ){
                #   Reverse selection: entries in `Output()`
                #   that have unnecessary layers parameters in 
                #   `OutputLayers`
                testOut <- uOutput2[, "Output()ID" ] %in% 
                    outputLayers[ selFixSelCol, "Output()ID" ] 
                
                message( sprintf( 
                    "Found %s entries in `OutputLayers` where selected is not set, for RUNID(s) %s", 
                    sum( selFixSelCol ), 
                    paste( unique( uOutput2[ testOut, "R_ID" ] ), collapse = "; " ) 
                ) ) 
                
                rm( testOut )
                
                #   Find the OutputLayerID to be removed
                idOut <- outputLayers[ selFixSelCol, "OutputLayerID" ]
                
                RODBC::sqlQuery( 
                    channel = channel, 
                    query   = sprintf( 
                        "UPDATE `OutputLayers` SET `Selected`=0 WHERE `OutputLayerID` IN (%s)", 
                        paste( as.character( idOut ), collapse = ", " ) 
                    ),  
                )   
                
                message( sprintf( 
                    "Set %s entries in `OutputLayers` (`Selected` set to 0)", 
                    length( idOut ) 
                ) ) 
                
                rm( idOut )
            }else{
                message( "Found no entries with `Selected` not set in `OutputLayers` (fine!)" )
            }   
            
            rm( selFixSelCol )
            
            
            
            # Close and exit
            # ----------------------------------------------
            
            RODBC::odbcClose( channel = channel ) 
            
            on.exit() 
        }   
    )   
    
    message( "Database cleaned" ) 
}   


